File: java/sfo_50_find_the_first_nonrepeating_character_in_a_string_s1/sfo_50_find_the_first_nonrepeating_character_in_a_string_s1.java
/*
* File: sfo_50_find_the_first_nonrepeating_character_in_a_string_s1.java
* Created Time: 2021-12-09
* Author: Krahets (krahets@163.com)
*/

package sfo_50_find_the_first_nonrepeating_character_in_a_string_s1;

import include.*;
import java.util.*;

// ===== Solution Code =====
class Solution {
    public char firstUniqChar(String s) {
        HashMap<Character, Boolean> dic = new HashMap<>();
        char[] sc = s.toCharArray();
        for(char c : sc)
            dic.put(c, !dic.containsKey(c));
        for(char c : sc)
            if(dic.get(c)) return c;
        return ' ';
    }
}

public class sfo_50_find_the_first_nonrepeating_character_in_a_string_s1 {
    public static void main(String[] args) {
        // ======= Test Case =======
        String s = "abaccdeff";
        // ====== Driver Code ======
        Solution slt = new Solution();
        char res = slt.firstUniqChar(s);
        System.out.println(res);
    }
}

File: sword_for_offer/codes/python/sfo_50_find_the_first_nonrepeating_character_in_a_string_s1.py
'''
File: sfo_50_find_the_first_non-repeating_character_in_a_string_s1.py
Created Time: 2021-12-09
Author: Krahets (krahets@163.com)
'''

from include import *

# ===== Solution Code =====
class Solution:
    def firstUniqChar(self, s: str) -> str:
        dic = {}
        for c in s:
            dic[c] = not c in dic
        for c in s:
            if dic[c]: return c
        return ' '

# ======= Test Case =======
s = "abaccdeff"
# ====== Driver Code ======
slt = Solution()
res = slt.firstUniqChar(s)
print(res)
